core/src/services/gcs/core.rs

```diff
@@ -267,6 +276,12 @@ impl GcsCore {
 
         let mut req = Request::post(&url);
 
+        if let Some(user_metadata) = op.user_metadata() {
+            for (key, value) in user_metadata {
+                req = req.header(format!("{X_GOOG_META_PREFIX}{key}"), value)
```

Comment:
Hi, this seems only apply to [when used the final request of a resumable upload](https://cloud.google.com/storage/docs/json_api/v1/parameters#xgoogmeta).

To set custom metadata, we need to use `insert`'s `uploadType=multipart`. Read https://cloud.google.com/storage/docs/uploading-objects 's `JSON API multipart upload (a single-request upload that includes object metadata)` for more information.

Reply:
In the details section of the first documentation you depicted, it says: 

> Any custom metadata set in the [initial request](https://cloud.google.com/storage/docs/performing-resumable-uploads#initiate-session) that uses the same KEY string is overwritten using the value provided in the X-Goog-Meta- header.

And in https://cloud.google.com/storage/docs/performing-resumable-uploads#initiate-session It seems that it is allowed in single chunks upload, as I think it is done here:
![image](https://github.com/user-attachments/assets/1c9eeadb-b174-4489-9158-a500cf0dae9d)


As you raised this, in multipart uploads I think we should include the headers in the `gcs_complete_multipart_upload` 
![image](https://github.com/user-attachments/assets/72d744c7-00f9-4de5-8ff0-fcfeaff77334)


In the behaviour tests are there test for both multipart and single chunk uploads with gcs? I would like to check that both `gcs_complete_multipart_upload` and `gcs_insert_object_request` set the metadata successfully.

Reply:
> As you raised this, in multipart uploads I think we should include the headers in the

Hi, the `multipart upload` is different from google's `uploadType=multipart`. Using `uploadType=multipart` inside `insert` means upload metadata with data at the same time via HTTP [Multipart Form](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST#multipart_form_submission). It's a different thing from s3's multipart upload which is in different requests.

Reply:
Hi, I understand what you mean. Include the user metadata in https://github.com/apache/opendal/blob/330bd74f3ace74f15966dfb2465191aa7c1ae5a6/core/src/services/gcs/core.rs#L303.

(and I also have to change https://github.com/apache/opendal/blob/330bd74f3ace74f15966dfb2465191aa7c1ae5a6/core/src/services/gcs/core.rs#L268 to `multipart/related`)

I have doubts about how this will behave if we mix _system_ metadata such as `storageClass` and `cacheControl` with user-defined metadata in that request.

I can't propertly test it thought... The command `OPENDAL_TEST=gcs cargo test behavior --features tests` is failing for me in local. It worked with `OPENDAL_TEST=azblob`...
![image](https://github.com/user-attachments/assets/7ef87990-3189-4018-94c7-6a4a88d43346)

Reply:
Hi, @jorgehermo9, the `metadata` json we sent to gcs will be like:

```json
{
  "contentType": "application/pdf",
  "metadata": {
      "location": "anywhere"
  }
}
```

As shown as https://cloud.google.com/storage/docs/json_api/v1/objects/insert#request-body.

So we are not mixing them with `user-defined metadata`.

Reply:
We can improve this part by writing this type directly without using `json!()`, if you prefer.

For example:

```rust
#[serde(default, rename_all = "camelCase")]
struct InsertRequestBody {
  #[serde(skip_serializing_if = "Option::is_none")]
  content_type: Option<String>,
  #[serde(skip_serializing_if = "HashMap::is_empty")]
  metadata: HashMap<String, String>
}
```

Reply:
Oh, I understand now. Crystal clear 😀 

It is my first time reading gcs documentation and I'm sorry for the inconveniences

thanks! I'll take a look into it

Reply:
I understand it better with that type. The hashmap that is currently in the code was confusing me!

Reply:
> It is my first time reading gcs documentation and I'm sorry for the inconveniences

Never mind. The GCS documentation is really poor; I've read it many times to find those details.

> I understand it better with that type. The hashmap that is currently in the code was confusing me!

Welcome to improve them!

Reply:
I think I addressed it at c2074104d55118645770a48e92a509c07ad5e548. I hope it looks more like what you meant.

Can you trigger CI and see how it breaks..?